### PR TITLE
Add baseline for robot framework and integration with UTF

### PR DESCRIPTION
**Pull request by lewijacn**
Signed-off-by: Tanner Lewis <lewijacn@amazon.com>

### Description
Introduces the "robot_tests" package which utilizes the robot framework for keyword driven BDD testing.
* Includes basic test for a document consistency expectation
* Includes basic Python client library for keyword setup

Integrates with UTF framework as an installed package to execute pre-upgrade and post-upgrade tests as steps in the given workflow

Several improvements to be made upon this (some that come to mind are below), but goal of this PR is to create a baseline to unblock others and be iterated upon
* Rework client library to use requests instead of python client
* Think through intermediate storage of test state in the robot_tests package
* Add additional robot tests to get clearer picture of the functionality here

### Issues Resolved
* https://github.com/opensearch-project/opensearch-migrations/issues/46

### Testing
Snapshot restore config
```
{
  "clusters_def": {
    "source": {
      "engine_version": "ES_7_10_2",
      "image": "docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2",
      "node_count": 2,
      "additional_node_config": {
        "ES_JAVA_OPTS": "-Xms512m -Xmx512m"
      }
    },
    "target": {
      "engine_version": "OS_1_3_6",
      "image": "opensearchproject/opensearch:1.3.6",
      "node_count": 2,
      "additional_node_config": {
        "plugins.security.disabled": "true",
        "OPENSEARCH_JAVA_OPTS": "-Xms512m -Xmx512m"
      }
    }
  },
  "upgrade_def": {
    "style": "snapshot-restore"
  }
}
```
Snapshot restore UTF log
```
(.venv) (05:36:07) local [~/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades] -> ./run_utf.py --test_config=test_config.json
[FrameworkRunner] Running Step: LoadTestConfig
[LoadTestConfig] Loading test config file...
[LoadTestConfig] Loaded test config file successfully
[FrameworkRunner] Step Succeeded: LoadTestConfig
[FrameworkRunner] Running Step: BootstrapDocker
[BootstrapDocker] Checking if Docker is installed and available...
[BootstrapDocker] Docker appears to be installed and available
[BootstrapDocker] Ensuring the Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available either locally or remotely...
[BootstrapDocker] Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available
[BootstrapDocker] Ensuring the Docker image opensearchproject/opensearch:1.3.6 is available either locally or remotely...
[BootstrapDocker] Docker image opensearchproject/opensearch:1.3.6 is available
[FrameworkRunner] Step Succeeded: BootstrapDocker
[FrameworkRunner] Running Step: SnapshotRestoreSetup
[SnapshotRestoreSetup] Creating shared Docker volume to share snapshots...
[SnapshotRestoreSetup] Created shared Docker volume cluster-snapshots-volume
[FrameworkRunner] Step Succeeded: SnapshotRestoreSetup
[FrameworkRunner] Running Step: ExpectationSelector
[ExpectationSelector] Querying applicable expectations...
[FrameworkRunner] Step Succeeded: ExpectationSelector
[FrameworkRunner] Running Step: StartSourceCluster
[StartSourceCluster] Creating source cluster...
[StartSourceCluster] Waiting up to 30 sec for cluster to be active...
Node source-cluster-node-2 is now active
Node source-cluster-node-1 is now active
[StartSourceCluster] Cluster source-cluster is active
[FrameworkRunner] Step Succeeded: StartSourceCluster
[FrameworkRunner] Running Step: TestSourceCluster
[TestSourceCluster] Querying cluster status...
[TestSourceCluster] ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role master name
172.30.0.2           11          67  55    0.85    0.45     0.41 dimr      *      source-cluster-node-1
172.30.0.3           30          67  47    0.85    0.45     0.41 dimr      -      source-cluster-node-2
[TestSourceCluster] Uploading sample document...
[TestSourceCluster] Retrieving uploaded document...
[TestSourceCluster] Retrieved uploaded doc sucessfully
[TestSourceCluster] {
    "_id": "1",
    "_index": "noldor",
    "_primary_term": 1,
    "_seq_no": 0,
    "_source": {
        "name": "Finwe"
    },
    "_type": "_doc",
    "_version": 1,
    "found": true
}
[FrameworkRunner] Step Succeeded: TestSourceCluster
[FrameworkRunner] Running Step: PreUpgradeRobotTest
==============================================================================
Tests                                                                         
==============================================================================
Tests.Common Upgrade Test                                                     
==============================================================================
Perform pre-upgrade setup of "consistent-document-count" expectation  PUT http://localhost:9200/sample-index [status:200 request:0.118s]
.POST http://localhost:9200/sample-index/_doc [status:201 request:0.115s]
.POST http://localhost:9200/sample-index/_doc [status:201 request:0.021s]
.POST http://localhost:9200/sample-index/_refresh [status:200 request:0.030s]
.POST http://localhost:9200/sample-index/_count [status:200 request:0.087s]
Perform pre-upgrade setup of "consistent-document-count" expectation  | PASS |
------------------------------------------------------------------------------
Tests.Common Upgrade Test                                             | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/pre_upgrade/output.xml
Log:     /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/pre_upgrade/log.html
Report:  /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/pre_upgrade/report.html
[FrameworkRunner] Step Succeeded: PreUpgradeRobotTest
[FrameworkRunner] Running Step: CreateSourceSnapshot
[CreateSourceSnapshot] Creating snapshot of source cluster...
[CreateSourceSnapshot] Confirming snapshot of source cluster exists...
[CreateSourceSnapshot] Source snapshot created successfully
[CreateSourceSnapshot] {
    "snapshots": [
        {
            "data_streams": [],
            "duration_in_millis": 0,
            "end_time": "1970-01-01T00:00:00.000Z",
            "end_time_in_millis": 0,
            "failures": [],
            "include_global_state": true,
            "indices": [
                "sample-index",
                "noldor"
            ],
            "shards": {
                "failed": 0,
                "successful": 0,
                "total": 0
            },
            "snapshot": "1",
            "start_time": "2022-12-20T22:36:30.302Z",
            "start_time_in_millis": 1671575790302,
            "state": "IN_PROGRESS",
            "uuid": "xT8KCksCT9uqtmbr4L7cHw",
            "version": "7.10.2",
            "version_id": 7100299
        }
    ]
}
[FrameworkRunner] Step Succeeded: CreateSourceSnapshot
[FrameworkRunner] Running Step: StartTargetCluster
[StartTargetCluster] Creating target cluster...
[StartTargetCluster] Waiting up to 30 sec for cluster to be active...
Node target-cluster-node-1 is now active
Node target-cluster-node-2 is now active
[StartTargetCluster] Cluster target-cluster is active
[FrameworkRunner] Step Succeeded: StartTargetCluster
[FrameworkRunner] Running Step: TestTargetCluster
[TestTargetCluster] ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role master name
172.31.0.2           42          93  58    1.98    0.73     0.50 dimr      -      target-cluster-node-1
172.31.0.3           41          93  55    1.98    0.73     0.50 dimr      *      target-cluster-node-2
[FrameworkRunner] Step Succeeded: TestTargetCluster
[FrameworkRunner] Running Step: RestoreSourceSnapshot
[RestoreSourceSnapshot] Checking if source snapshots are visible on target...
[RestoreSourceSnapshot] Source snapshot visible to target cluster
[RestoreSourceSnapshot] Restoring source snapshot onto target...
[RestoreSourceSnapshot] Waiting a few seconds for the snapshot to be restored...
[RestoreSourceSnapshot] Attempting to retrieve source doc from target cluster...
[RestoreSourceSnapshot] Document retrieved, snapshot restored successfully
[RestoreSourceSnapshot] {
    "_id": "1",
    "_index": "noldor",
    "_primary_term": 1,
    "_seq_no": 0,
    "_source": {
        "name": "Finwe"
    },
    "_type": "_doc",
    "_version": 1,
    "found": true
}
[FrameworkRunner] Step Succeeded: RestoreSourceSnapshot
[FrameworkRunner] Running Step: PostUpgradeRobotTest
==============================================================================
Tests                                                                         
==============================================================================
Tests.Common Upgrade Test                                                     
==============================================================================
Perform post-upgrade assertion of "consistent-document-count" expe... .POST http://localhost:9202/sample-index/_count [status:200 request:0.104s]
Perform post-upgrade assertion of "consistent-document-count" expe... | PASS |
------------------------------------------------------------------------------
Tests.Common Upgrade Test                                             | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/post_upgrade/output.xml
Log:     /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/post_upgrade/log.html
Report:  /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades/robot_results/post_upgrade/report.html
[FrameworkRunner] Step Succeeded: PostUpgradeRobotTest
[FrameworkRunner] Running Step: StopSourceCluster
[StopSourceCluster] Stopping cluster source-cluster...
[StopSourceCluster] Cleaning up underlying resources for cluster source-cluster...
[FrameworkRunner] Step Succeeded: StopSourceCluster
[FrameworkRunner] Running Step: StopTargetCluster
[StopTargetCluster] Stopping cluster target-cluster...
[StopTargetCluster] Cleaning up underlying resources for cluster target-cluster...
[FrameworkRunner] Step Succeeded: StopTargetCluster
[FrameworkRunner] Running Step: SnapshotRestoreTeardown
[SnapshotRestoreTeardown] Removing shared Docker volume cluster-snapshots-volume...
[SnapshotRestoreTeardown] Removed shared Docker volume cluster-snapshots-volume
[FrameworkRunner] Step Succeeded: SnapshotRestoreTeardown
[FrameworkRunner] Ran through all steps successfully
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-20_17_36_16
```
### Check List
- [ ] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: upgrades/upgrade_testing_framework/steps/step_expectation_selector.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+    def _run(self):
+        # Get the state we need
+
+        # Begin the step body
+        # Query the expectation knowledge base
+        self.logger.info("Querying applicable expectations...")
+
+        # Update our state
+        self.state.eligible_expectations = ['consistent-document-count']
```
Comment (mikaylathompson) [inline upgrades/upgrade_testing_framework/steps/step_expectation_selector.py:16]: minor, but maybe worth adding a TODO-type comment that this is a temporary measure

File: robot_tests/robot_tests/test_executor.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+package_directory = os.path.dirname(os.path.abspath(__file__))
+
+
+class TestExecutor:
+    def __init__(self, host: str, port: int):
+        self.host = host
+        self.port = port
+
+    def execute_tests(self, include_tags=[], exclude_tags=[], output_dir="robot_results"):
+        run(f"{package_directory}/tests", include=include_tags,exclude=exclude_tags, outputdir=output_dir,
```
Comment (mikaylathompson) [inline robot_tests/robot_tests/test_executor.py:12]: no need to do this now, but is it possible for us to get test results/other data back from this method? It would make it very easy to store info in the utf `state` to give it to the reporter step

**Review by mikaylathompson — APPROVED** [review]

File: robot_tests/robot_tests/test_executor.py (hunk 2 above)
Comment (chelma) [reply robot_tests/robot_tests/test_executor.py:12]: Yeah, I had the same thought; curious if Tanner has any ideas.  I'll see if anything pops into mind once I read through the rest of the PR as well (need to see how some things are structured).

File: robot_tests/robot_tests/test_executor.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+from robot import run
+package_directory = os.path.dirname(os.path.abspath(__file__))
+
+
+class TestExecutor:
+    def __init__(self, host: str, port: int):
+        self.host = host
+        self.port = port
+
+    def execute_tests(self, include_tags=[], exclude_tags=[], output_dir="robot_results"):
```
Comment (chelma) [inline robot_tests/robot_tests/test_executor.py:11]: The output should be in the UTF workspace directory; see the `upgrade_testing_framework.core.workspace_wrangler` for more details about that.

File: robot_tests/robot_tests/tests/common_upgrade_test.robot [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    Refresh index  sample-index
+    ${count} =  Count documents in index  sample-index
+    Store data with label  ${count}  consistent-document-count
+    Should Be Equal  ${count}  ${2}
+
+Perform post-upgrade assertion of "consistent-document-count" expectation
+    [Tags]  consistent-document-count  post-upgrade
+    ${stored_count} =  Retrieve stored number by label  consistent-document-count
+    ${count} =  Count documents in index  sample-index
+    Should Be Equal  ${stored_count}  ${count}
```
Comment (chelma) [inline robot_tests/robot_tests/tests/common_upgrade_test.robot:26]: I like where this is going, a lot.  Really cool.  Shoutout to @sumobrian for the recommendation of this framework!

File: robot_tests/setup.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import setuptools
+
+setuptools.setup(
+    name="robottests",
```
Comment (chelma) [inline robot_tests/setup.py:4]: We should make the Python package name specific to our use-case; `robottests` could apply to anyone's robot framework test implementation.  How about `upgrade_testing_framework_robot`, or something?

File: upgrades/upgrade_testing_framework/steps/__init__.py [hunk 6]
```diff
@@ -11,3 +11,6 @@
 from upgrade_testing_framework.steps.step_stop_target_cluster import StopTargetCluster
 from upgrade_testing_framework.steps.step_test_source_cluster import TestSourceCluster
 from upgrade_testing_framework.steps.step_test_target_cluster import TestTargetCluster
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/__init__.py:13]: Probably don't need these any more; how do you feel about removing them?

File: upgrades/upgrade_testing_framework/steps/step_expectation_selector.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+from upgrade_testing_framework.core.framework_step import FrameworkStep
+
+class ExpectationSelector(FrameworkStep):
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_expectation_selector.py:3]: In general, I'd prefer we name our FrameworkSteps as an action-to-be-performed rather than a thing-doing-work.  So, how about `SelectExpectations` instead?  The idea is the high-level log output should be as descriptive as possible and the step name is a key part of understanding that.

File: upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+from upgrade_testing_framework.core.framework_step import FrameworkStep
+from robot_tests.test_executor import TestExecutor
+
+class PostUpgradeRobotTest(FrameworkStep):
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py:4]: Same comment; how about `PerformPostUpgradeTest` instead?  Not sure we need to include "Robot" in the step name.

File: upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+    def _run(self):
+        # Get the state we need
+        target_cluster = self.state.target_cluster
+
+        # Pull host and credentials (if security is enabled) from state
+        test_executor = TestExecutor("localhost", target_cluster.rest_ports[0])
+
+        # Begin the step body
+        test_executor.execute_tests(include_tags=self.state.eligible_expectations,
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py:17]: This one is pretty important to me - please follow the FrameworkStep convention of always pulling state into a local variable in the dedicated section at the beginning of the `_run()` definition.  As FrameworkSteps get more complex, this will keep us from pulling out hair out.  The state is basically a global variable and we should treat it with a combination of reverence/revulsion and keep it in a very explicit box.

Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py:17]: Similarly, please never write to the global state outside the prescribed section at the end of a FrameworkStep

File: upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
+from upgrade_testing_framework.core.framework_step import FrameworkStep
+from robot_tests.test_executor import TestExecutor
+
+class PreUpgradeRobotTest(FrameworkStep):
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py:4]: Same comment; `PerformPreUpgradeTests`

File: upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+    def _run(self):
+        # Get the state we need
+        source_cluster = self.state.source_cluster
+
+        # Pull host and credentials (if security is enabled) from state
+        test_executor = TestExecutor("localhost", source_cluster.rest_ports[0])
+
+        # Begin the step body
+        test_executor.execute_tests(include_tags=self.state.eligible_expectations,
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py:17]: Same comment - only pull state in the prescribed section of the `_run()` implementation

File: upgrades/upgrade_testing_framework/steps/step_expectation_selector.py (hunk 1 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_expectation_selector.py:16]: Please add the `eligible_expectations` to the FrameworkState as a declared member.  It seems like it will be here to stay, so it should be a first-class citizen, State-wise.

**Review by chelma — COMMENTED** [review]
A few high-level comments
* We need to do some pruning of the messaging that gets send to STDOUT, as the "high level" output is getting to the point of being hard to parse.  I have some ideas on how to make things more readable that I can quickly roll into my current task to make the REST client library.  Also, I never intended for us to print the REST responses for things like the snapshot/restore calls long term and just had those for demo purposes.  Removing them will improve things quite a bit too.
* Now that we have actual tests we can run, it's unclear whether the `TestSourceCluster` and `TestTargetCluster` steps add any real value.  I had intended for them to go the way of all mortal code once the Expectations were on line (ashes to ashes, dust to dust), but curious to hear y'all's thoughts.

File: robot_tests/robot_tests/tests/common_upgrade_test.robot [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+#    [Arguments]    ${index_name}
+#    Create index   ${index_name}
+
+*** Test Cases ***
+
+# Link to expectation description
+Perform pre-upgrade setup of "consistent-document-count" expectation
+    [Tags]  consistent-document-count  pre-upgrade
+    Create index  sample-index
+    Create document in index  sample-index  {"color":"blue", "name":"sky"}
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply robot_tests/robot_tests/tests/common_upgrade_test.robot:15]: Is the `keywords.resource` file a way to group some of these common commands together under a new string of keywords? If so I'm definitely onboard for that

File: upgrades/upgrade_testing_framework/steps/step_expectation_selector.py (hunk 1 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_expectation_selector.py:16]: Agreed for both will update in next revision

File: robot_tests/robot_tests/test_executor.py (hunk 3 above)
Comment (lewijacn) [reply robot_tests/robot_tests/test_executor.py:11]: Cool needed a better place to put this, will add in

File: robot_tests/setup.py (hunk 5 above)
Comment (lewijacn) [reply robot_tests/setup.py:4]: My thinking here is that I don't see the robot tests being directly tied to the UTF. No idea if we want to go this direction but I could potentially see something like an "assessment" tag being added to a set of tests or someone running certain tests on their own cluster. Will agree this package name is a bit vague though and maybe something like `robot_cluster_tests` or `robot_upgrade_tests` (even this is a bit leading) is better

File: upgrades/upgrade_testing_framework/steps/__init__.py (hunk 6 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/__init__.py:13]: Do you think you have any need for validating a cluster has started properly with these steps? I was originally thinking you might do some startup validation here or any checks that couldn't be accomplished through the API

File: upgrades/upgrade_testing_framework/steps/step_expectation_selector.py (hunk 7 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_expectation_selector.py:3]: Agreed will add

File: upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py (hunk 8 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py:4]: Agreed I was on the fence about adding robot myself, will modify

File: upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py (hunk 9 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_post_upgrade_robot_test.py:17]: Good thought will add in

File: upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py (hunk 10 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py:4]: Agreed will modify

File: upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py (hunk 11 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_pre_upgrade_robot_test.py:17]: Agreed will modify

File: robot_tests/setup.py (hunk 5 above)
Comment (chelma) [reply robot_tests/setup.py:4]: I'm fine w/ whatever name as long as it is more specific to what it's doing.  `robot_tests_cluster_upgrade`?

File: upgrades/upgrade_testing_framework/steps/__init__.py (hunk 6 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/__init__.py:13]: Good question!  Ideally, the test that stands up the clusters should not exit successfully unless they're properly started, yeah?  In that case, we still don't need these `Test*` steps.

File: robot_tests/robot_tests/test_executor.py (hunk 2 above)
Comment (lewijacn) [reply robot_tests/robot_tests/test_executor.py:12]: Good idea. I think we could pretty easily store the location of the output files (Which should be placed in the UTF workspace dir as Chris mentioned) from these tests in the `utf` state (Not necessarily as a return of this function) and then output this in the reporter step. I think this is all I had in mind at this point but open to other thoughts

File: upgrades/upgrade_testing_framework/steps/__init__.py (hunk 6 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/__init__.py:13]: Hey, actually, changed my mind about these.  Let me remove these later, as I have some in-flight changes that will get messy if you remove them.

**Comment by lewijacn** [comment]
> A few high-level comments
> 
> * We need to do some pruning of the messaging that gets send to STDOUT, as the "high level" output is getting to the point of being hard to parse.  I have some ideas on how to make things more readable that I can quickly roll into my current task to make the REST client library.  Also, I never intended for us to print the REST responses for things like the snapshot/restore calls long term and just had those for demo purposes.  Removing them will improve things quite a bit too.
> * Now that we have actual tests we can run, it's unclear whether the `TestSourceCluster` and `TestTargetCluster` steps add any real value.  I had intended for them to go the way of all mortal code once the Expectations were on line (ashes to ashes, dust to dust), but curious to hear y'all's thoughts.

I think we're clear now on second point, will remove soon in another PR. For the first point I agree, currently we are just logging everything for the robot framework and will be a bit overwhelming as this grows. I plan to have a future change to minimize this and direct users to the created report by default.

File: upgrades/upgrade_testing_framework/core/framework_state.py [hunk 13]
```diff
@@ -21,6 +21,7 @@ def __init__(self, state: dict):
         self.source_cluster: Cluster = None
         self.target_cluster: Cluster = None
         self.test_config: TestConfig = None
+        self.eligible_expectations = []
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/core/framework_state.py:24]: Not a blocker, but we should indicate what type is contained within this list.

File: upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py [hunk 14]
```diff
@@ -1,22 +1,25 @@
...
     """
     This step is where you run post-upgrade tests
     """
 
     def _run(self):
         # Get the state we need
         target_cluster = self.state.target_cluster
+        port = target_cluster.rest_ports[0]
+        expectations = self.state.eligible_expectations
+        output_directory = f"{self.state.get_key('test_results_directory')}/post_upgrade"
```
Comment (chelma) [inline upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py:14]: Also not a blocker for this PR, but I would propose that this section of a `_run()` body be used *solely* to pull global state into local variables, without any changes or modifications.  The premise is to make an incredibly clear boundary between the FrameworkStep and he global state it interacts with.

Going forward, I'd like to impose this as a non-negotiable convention.  @mikaylathompson, FYI.

**Review by chelma — APPROVED** [review]
Approved w/ two non-blocking comments.

File: upgrades/upgrade_testing_framework/core/framework_state.py (hunk 13 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/core/framework_state.py:24]: Wasn't sure if you could do this with Python or not, will update this with next PR

File: upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py (hunk 14 above)
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py:14]: A little inconvenient but probably a good convention to have as this does create some ambiguity with what the state file has versus what was actually used. Will update with next PR

File: upgrades/upgrade_testing_framework/core/framework_state.py (hunk 13 above)
Comment (mikaylathompson) [reply upgrades/upgrade_testing_framework/core/framework_state.py:24]: fine to have this in the next PR, but for future reference, you can do 
```
from typing import List
self.eligible_expectations: List[str] = []
```

File: upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py (hunk 14 above)
Comment (mikaylathompson) [reply upgrades/upgrade_testing_framework/steps/step_perform_post_upgrade_test.py:14]: I agree with that as a convention. I wonder if there's a good way to enforce/check for it down the road, but for now, agreed that we should self-impose it.
